posts/2018-06-07-shared-state-in-fp.md

```diff
@@ -0,0 +1,372 @@
+---
+layout: post
+title: Shared State in Functional Programming
+
+meta:
+  nav: blog
+  author: gvolpe
+  pygments: true
+
+tut:
+  scala: 2.12.4
+  binaryScala: "2.12"
+  dependencies:
+    - org.scala-lang:scala-library:2.12.4
+    - org.typelevel::cats-core:1.1.0
+
+---
+
+Newcomers to Functional Programming are often very confused about the proper way to share state without breaking purity and end up having a mix of pure and impure code that defeats the purpose of having pure FP code in the first place.
+
+This is the reason that has motivated me to write a beginner friendly guide :)
+
+## Referential Transparency
+
+I think this is the key to understanding the benefits of Functional Programming. Without it, it'll be hard to understand other deeper concepts, such as `Shared State`. So let's start this guide by refreshing our knowledge about it:
+
+##### What is Referential Transparency?
+
+Being able to substitute an expression by its bound value without changing the meaning of the program.
+
+##### What does this mean?
+
+An example is worth thousand words. Consider this:
+
+```scala
+val expr = 123
+(expr, expr)
+```
+
+If we now substitute the expression by its bound value, does the meaning of the program change?
+
+```
```

Comment:
Consider using `tut`

Reply:
I'm :-1: on using `tut` for a blog post. I think it's great for docs to keep it updated but not really for a blog post. 

The reason being that to preview this blog post I must compile all the existent blog posts and we might have divergence in the future when updating dependencies and an old post that was completely valid back in time is no longer valid, even if we force it to compile, because also the sentences will need to be updated.